### ISSUE TITLE
Rewrite all database update operations using transactions

### ISSUE THREAD

**Issue by tskir** [EBIvariation/trait-curation#52]
When initiating an update from ClinVar or from ZOOMA, some exception can occur in the process.

To avoid leaving the database in an inconsistent state, all such operations should proceed in approximately this manner:
* Lock database for write operations
* Start transaction
* Attempt import
  + If success: commit transaction, unlock database
  + In case of any exception: roll back transaction, unlock database, log an error

**Comment by joj0s** [comment]
So the way I am thinking of implementing this, is to use a combination of F expressions (https://docs.djangoproject.com/en/2.2/ref/models/expressions/#avoiding-race-conditions-using-f) and transaction decorators (https://docs.djangoproject.com/en/3.0/topics/db/transactions/#transactions), what do you think?

**Comment by joj0s** [comment]
Or actually, this might be a better choice to apply locks https://docs.djangoproject.com/en/3.0/ref/models/querysets/#select-for-update

**Comment by tskir** [comment]
At least on the surface, it looks to me that the _F expressions_ and _select_for_update_ are useful for doing single operations, like changing some fields. Since importing from ClinVar/ZOOMA/OLS affects a number of entities in a non-trivial way, I think we should have a proper transaction, explicitly locking _everything_ and rolling back _everything_ for exception. So I think we should go with transactions.

Looking at the docs, it looks to me that [transaction decorators](https://docs.djangoproject.com/en/3.0/topics/db/transactions/#controlling-transactions-explicitly) do just what we need, in the most elegant way possible:
```python
from django.db import transaction

@transaction.atomic
def viewfunc(request):
    # This code executes inside a transaction.
    do_stuff()
```

**Comment by tskir** [comment]
Locking down write access to the database before starting a transaction is a separate issue which will have to be researched and discussed separately. I'm not sure how to do this best. Perhaps we can set up some sort of an app state variable which would describe whether user modifications to the data should be allowed. Then, each function which starts the transaction could also explicitly set this variable to False while starting, and return to True after everything has been successfully completed. Perhaps there are more elegant ways of doing this.
